functions/test/github.test.js

```diff
@@ -1,121 +1,223 @@
+import {
+  addAndGetSnapshot,
+  setAndGetSnapshot,
+  setupDbAdmin,
+  teardownDb
+} from './helpers'
+
+const sha1 = require('sha1')
 const nock = require('nock')
 const fs = require('fs')
 const path = require('path')
 
-const Octokit = require('@octokit/rest')
-
-// Requiring our app implementation
 const Github = require('../lib/github')
+const prOpenedPayload = require('./fixtures/pull_request.opened')
 
-// Requiring our fixtures
-const payload = require('./fixtures/pull_request.opened')
-const hugePayload = require('./fixtures/pull_request.opened.huge')
-const commits = require('./fixtures/commits')
+const contributionKey = 'github.com/bocon13/cla-test/pull/3'
+const contributionId = sha1(contributionKey)
+const identity = 'github:bocon13'
 
-const checkIdentifiesSuccess = async () => {
-  return Promise.resolve({
-    allWhitelisted: true,
-    missingIdentities: []
-  })
-}
+const githubApi = 'https://api.github.com'
+const statusUri = '/repos/bocon13/cla-test/statuses/2129e453f6d652badfb353c510a3669873a15f7c'
+const createCommentUri = '/repos/bocon13/cla-test/issues/3/comments'
+const existingCommentUri = '/repos/bocon13/cla-test/issues/comments/1'
 
-const checkIdentifiesFail = async () => {
-  return Promise.resolve({
-    allWhitelisted: false,
-    missingIdentities: []
-  })
-}
+const appId = 123
+const projectId = 'github-test-' + new Date()
 
 nock.disableNetConnect()
 
-describe('Github PR Webhook', () => {
-  const appId = 123
+describe('Github lib', () => {
+  let db
   let github
+  let contribsRef
+  let eventsRef
+  let whitelistsRef
   let mockCert
-  const cla = {
-    checkIdentities: checkIdentifiesSuccess
-  }
+  let mockContribution
+  let mockEvent
+  let payload
 
-  beforeAll((done) => {
+  beforeAll(async (done) => {
     fs.readFile(path.join(__dirname, 'fixtures/mock-cert.pem'), (err, cert) => {
       if (err) return done(err)
-      mockCert = cert
+      mockCert = cert.toString()
       return done()
     })
   })
 
-  beforeEach(() => {
-    github = new Github(appId, mockCert, 'secret', cla)
+  beforeEach(async () => {
+    db = await setupDbAdmin(null, projectId)
+    contribsRef = db.collection('contributions')
+    eventsRef = db.collection('events')
+    whitelistsRef = db.collection('whitelists')
+    github = new Github(appId, mockCert, 'secret', db)
     // Return a token for fake application (id=555)
-    nock('https://api.github.com')
-      .post('/app/installations/555/access_tokens')
-      .reply(200, { token: 'test' })
-    // Return the list of commits for PR #3
-    nock('https://api.github.com')
-      .get('/repos/bocon13/cla-test/pulls/3/commits')
-      .reply(200, commits)
+    nock(githubApi).post('/app/installations/555/access_tokens')
+      .reply(201, { token: 'test' })
+    // Make a copy so we can modify it in each test
+    payload = JSON.parse(JSON.stringify(prOpenedPayload))
```

Comment:
why are we parsing a JSON that we're converting to a string? We'll end up with the same value we started with

Reply:
I want to make a deep copy of the original JSON such that each test case can modify it without affecting other test cases. Is there a better way of doing that?